### PR TITLE
cop: Migrate implementation and test for Rpn Function DateFormatSig

### PR DESCRIPTION
**Pull request by mapleFU**
## What have you changed? (mandatory)

Add implementation and tests for rpn function DateFormatSig.

You should pay attention to the files below:
* `src/coprocessor/dag/rpn_expr/impl_time.rs` I created this file and add implementation and tests for rpn function `DateFormatSig` here. 

In the original implementation, the function required `DateTime` and `&str`, the output data type is `Cow<'a, [u8]>`. And in my implementation, I use `Bytes` instead of `&str` and `Cow<'a, [u8]>`.

And because of the issue https://github.com/tikv/tikv/issues/3953 , I left a **TODO** and a commented test for testing it in the function `test_date_format_null` in `impl_time.rs`, please take care.

## What are the type of the changes? (mandatory)

New feature (change which adds functionality)

## How has this PR been tested? (mandatory)

New unit tests.

### CODE REVIEW

File: src/coprocessor/dag/rpn_expr/impl_time.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
+use cop_codegen::rpn_fn;
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_time.rs]: Missing license header.

File: src/coprocessor/dag/rpn_expr/impl_time.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    t: &Option<DateTime>,
+    layout: &Option<Bytes>,
+) -> Result<Option<Bytes>> {
+    use std::str::from_utf8;
+
+    if t.is_none() || layout.is_none() {
+        return Ok(None);
+    }
+    let (t, layout) = (t.as_ref().unwrap(), layout.as_ref().unwrap());
+    if t.invalid_zero() {
```
Comment (AndreMouche) [inline src/coprocessor/dag/rpn_expr/impl_time.rs]: Any test case cover it?

File: src/coprocessor/dag/rpn_expr/impl_time.rs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    ctx: &mut EvalContext,
+    t: &Option<DateTime>,
+    layout: &Option<Bytes>,
+) -> Result<Option<Bytes>> {
+    use std::str::from_utf8;
+
+    if t.is_none() || layout.is_none() {
+        return Ok(None);
+    }
+    let (t, layout) = (t.as_ref().unwrap(), layout.as_ref().unwrap());
```
Comment (AndreMouche) [inline src/coprocessor/dag/rpn_expr/impl_time.rs]: What will `mysql` do when `t` is an invalid zero and layout is none?

Comment (mapleFU) [reply src/coprocessor/dag/rpn_expr/impl_time.rs]: > What will `mysql` do when `t` is an invalid zero and layout is none?

It's layout is none both in and not in strict mode.

Comment (AndreMouche) [reply src/coprocessor/dag/rpn_expr/impl_time.rs]:
```
mysql> select date_format("0000-00-00","%Y");
+--------------------------------+
| date_format("0000-00-00","%Y") |
+--------------------------------+
| NULL                           |
+--------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+----------------------------------------+
| Level   | Code | Message                                |
+---------+------+----------------------------------------+
| Warning | 1292 | Incorrect datetime value: '0000-00-00' |
+---------+------+----------------------------------------+
1 row in set (0.00 sec)

mysql> select  @@session.sql_mode;
+-----------------------------------------------------------------------------------------------------------------------+
| @@session.sql_mode                                                                                                    |
+-----------------------------------------------------------------------------------------------------------------------+
| ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION |
+-----------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

Comment (mapleFU) [reply src/coprocessor/dag/rpn_expr/impl_time.rs]: Thanks for your review. On account of the issue https://github.com/tikv/tikv/issues/3953 , test for datetime '0000-00-00' will not be test now.

As for the warnings, if `t.invalid_zero()` works well, `EvalContext.handle_invalid_time_error` will be called, which will check the `sql_mode` of the system and show warning if necessary.

File: src/coprocessor/dag/rpn_expr/impl_time.rs (hunk 2 above)
Comment (mapleFU) [reply src/coprocessor/dag/rpn_expr/impl_time.rs]: Tests about None are added in latest commits.

File: src/coprocessor/dag/rpn_expr/impl_time.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                .unwrap();
+            assert_eq!(output, expect, "{:?} {:?}", date, format);
+        }
+    }
+
+    #[test]
+    fn test_date_format_null() {
+        // test date format when format is None
+        let cases: Vec<(&str, Option<&str>)> = vec![
+            ("2010-01-07 23:12:34.12345", None),
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_time.rs]: Why not use `Some("....")` which is more clear?

File: src/coprocessor/dag/rpn_expr/impl_time.rs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            //                            %x %Y %y %%",
+            //                ),
+            //            ),
+        ];
+
+        for (date, format) in cases {
+            let date = if date.is_empty() {
+                None
+            } else {
+                Some(DateTime::parse_utc_datetime(date, 6).unwrap())
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_time.rs]: You are also testing the parsing function. Better to construct zero date time directly.

File: src/coprocessor/dag/rpn_expr/impl_time.rs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                .push_param(date.clone())
+                .push_param(format.clone())
+                .evaluate(ScalarFuncSig::DateFormatSig)
+                .unwrap();
+            assert_eq!(output, expect, "{:?} {:?}", date, format);
+        }
+    }
+
+    #[test]
+    fn test_date_format_null() {
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_time.rs]: Looks like merging into one function is enough. There is no function called `date_format_null`.

File: src/coprocessor/dag/rpn_expr/impl_time.rs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+            let expect = Some(expect.as_bytes().to_vec());
+
+            let output = RpnFnScalarEvaluator::new()
+                .push_param(date.clone())
+                .push_param(format.clone())
+                .evaluate(ScalarFuncSig::DateFormatSig)
+                .unwrap();
+            assert_eq!(output, expect, "{:?} {:?}", date, format);
+        }
+    }
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_time.rs]: What about these cases in the original implementation?

```rust
        // test NULL case
        test_err_case_two_arg(
            &mut ctx,
            ScalarFuncSig::DateFormatSig,
            Datum::Null,
            Datum::Null,
        );
        // test zero case
        let mut cfg = EvalConfig::new();
        cfg.set_flag(Flag::IN_UPDATE_OR_DELETE_STMT)
            .set_sql_mode(SqlMode::ERROR_FOR_DIVISION_BY_ZERO | SqlMode::STRICT_ALL_TABLES);
        ctx = EvalContext::new(Arc::new(cfg));
        test_err_case_two_arg(
            &mut ctx,
            ScalarFuncSig::DateFormatSig,
            Datum::Null,
            Datum::Null,
        );
```

File: src/coprocessor/dag/rpn_expr/impl_time.rs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+) -> Result<Option<Bytes>> {
+    use std::str::from_utf8;
+
+    if t.is_none() || layout.is_none() {
+        return Ok(None);
+    }
+    let (t, layout) = (t.as_ref().unwrap(), layout.as_ref().unwrap());
+    if t.invalid_zero() {
+        return ctx
+            .handle_invalid_time_error(Error::incorrect_datetime_value(&format!("{}", t)))
```
Comment (lonng) [inline src/coprocessor/dag/rpn_expr/impl_time.rs]: Maybe we should add some tests to cover the invalid time situation in different SQL mode.

**Comment by breeswish** [comment]
@mapleFU Any updates?

File: src/coprocessor/dag/rpn_expr/impl_time.rs (hunk 7 above)
Comment (mapleFU) [reply src/coprocessor/dag/rpn_expr/impl_time.rs]: The test below `// test NULL case ` is covered in the `test_date_format_null` (I will update its name later).

File: src/coprocessor/dag/rpn_expr/impl_time.rs (hunk 8 above)
Comment (mapleFU) [reply src/coprocessor/dag/rpn_expr/impl_time.rs]: > Maybe we should add some tests to cover the invalid time situation in different SQL mode.

Now I add tests for strict SqlMode, but on account of #3953, it cannot be applied yet.

**Comment by mapleFU** [comment]
> @mapleFU Any updates?

Yes, PTAL.

**Comment by breeswish** [comment]
Where is the identical test for:

```rust
        // test zero case
        let mut cfg = EvalConfig::new();
        cfg.set_flag(Flag::IN_UPDATE_OR_DELETE_STMT)
            .set_sql_mode(SqlMode::ERROR_FOR_DIVISION_BY_ZERO | SqlMode::STRICT_ALL_TABLES);
        ctx = EvalContext::new(Arc::new(cfg));
        test_err_case_two_arg(
            &mut ctx,
            ScalarFuncSig::DateFormatSig,
            Datum::Null,
            Datum::Null,
        );
```

File: src/coprocessor/dag/rpn_expr/impl_time.rs [hunk 9]
```diff
@@ -97,28 +101,32 @@ mod tests {
...
+        let cases: Vec<(Option<&str>, Option<&str>)> = vec![
+            //            // TODO: pass this test after refactoring the issue #3953 is fixed.
+            //            (
+            //                Some("abc%b %M %m %c %D %d %e %j"),
+            //                Some("0000-00-00 00:00:00"),
+            //            ),
+            (None, None),
+        ];
+
+        for (format, time) in cases {
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_time.rs]: No need to be a loop for the None -> None case. Other test cases are not suitable for these flags.

Comment (mapleFU) [reply src/coprocessor/dag/rpn_expr/impl_time.rs]: > No need to be a loop for the None -> None case. Other test cases are not suitable for these flags.

What about test for `"0000-00-00 00:00:00"`?

File: src/coprocessor/dag/rpn_expr/impl_time.rs [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+            let format: Option<Bytes> = format.map(|s| s.as_bytes().to_vec());
+            let time: Option<DateTime> = time.map(|d| DateTime::parse_utc_datetime(d, 6).unwrap());
+
+            let mut cfg = EvalConfig::new();
+            cfg.set_flag(Flag::IN_UPDATE_OR_DELETE_STMT)
+                .set_sql_mode(SqlMode::ERROR_FOR_DIVISION_BY_ZERO | SqlMode::STRICT_ALL_TABLES);
```
Comment (lonng) [inline src/coprocessor/dag/rpn_expr/impl_time.rs]: Do we need the `SqlMode::ERROR_FOR_DIVISION_BY_ZERO`?

**Comment by lonng** [comment]
LGTM

File: src/coprocessor/dag/rpn_expr/impl_time.rs (hunk 10 above)
Comment (mapleFU) [reply src/coprocessor/dag/rpn_expr/impl_time.rs]: > Do we need the `SqlMode::ERROR_FOR_DIVISION_BY_ZERO`?

I migrated it from `builtin_time.rs`. And seems it doesn't make sense here. I will change it to `SqlMode::NO_ZERO_DATE | SqlMode::STRICT_ALL_TABLES`

File: src/coprocessor/dag/rpn_expr/impl_time.rs (hunk 9 above)
Comment (mapleFU) [reply src/coprocessor/dag/rpn_expr/impl_time.rs]: > No need to be a loop for the None -> None case. Other test cases are not suitable for these flags.

Oh, I just made a mistake. `t.invalid_zero` should return error in strict mode.
The tidb code is here: https://github.com/pingcap/tidb/blob/a737d26e4f126cbd6cff6872e10003c66280d423/expression/errors.go

File: src/coprocessor/dag/rpn_expr/impl_time.rs [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+            let time: Option<DateTime> = None;
+
+            let mut cfg = EvalConfig::new();
+            cfg.set_flag(Flag::IN_UPDATE_OR_DELETE_STMT)
+                .set_sql_mode(SqlMode::NO_ZERO_DATE | SqlMode::STRICT_ALL_TABLES);
+            let ctx = EvalContext::new(Arc::new(cfg));
+
+            let output = RpnFnScalarEvaluator::new()
+                .context(ctx)
+                .push_param(time.clone())
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_time.rs]: Why not simply `.push_param(None)`

File: src/coprocessor/dag/rpn_expr/impl_time.rs [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+                .push_param(format)
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_time.rs]: `.push_param(None)`

File: src/coprocessor/dag/rpn_expr/impl_time.rs [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+                .evaluate::<Bytes>(ScalarFuncSig::DateFormatSig)
+                .unwrap();
+            assert_eq!(output, None, "{:?} {:?}", time, "");
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_time.rs]: `assert_eq!(output, None)`

**Comment by breeswish** [comment]
CI failed

File: src/coprocessor/dag/rpn_expr/impl_time.rs (hunk 11 above)
Comment (kennytm) [reply src/coprocessor/dag/rpn_expr/impl_time.rs]: `Option::<X>::None` could be written as `None::<X>` instead.

**Pull request merged**